### PR TITLE
Make this repo support multiple architectures

### PR DESCRIPTION
**Pull request by luxas**
Fixes https://github.com/kubernetes/heapster/issues/1280

Written in line with proposal https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/multi-platform.md

Originally in https://github.com/luxas/kubernetes-on-arm/tree/master/addons/images

**Disclaimer**: I have not manually tested this yet.

I would appreciate feedback.
I left all amd64 code totally unchanged, but eventually I think we'd like to make the amd64 images smaller as well.

We could for example release a v1.2.1 with no code changes, just release infra changes if we want to

@piosz @vishh @brendandburns @ixdy @kubernetes/heapster-maintainers

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/heapster/1387)

### CODE REVIEW

**Comment by vishh** [comment]
🤗 👍

**Comment by luxas** [comment]
@vishh @kubernetes/heapster-maintainers This PR is now ready for review.

I'd like to get it merged soon, as we're planning to add the heapster deployment by default to kubeadm.

It's easy for you to just merge this and push the influxdb and grafana containers.
But we can discuss what we should do with the heapster image.
I think we have two options:
a) You checkout the v1.2.0 Go code, but use this releasing/deployment code when pushing the image
b) After this PR is merged, I'll send a PR that targets the `release-1.2` branch, and you can then release `v1.2.1` with that commit.

Also, I'd like to know if you'd like that the `amd64` images stay the same, or if they should be swapped to this "new" code (that's risky, but you choose)

Please take a look as soon as possible.

**Comment by piosz** [comment]
@luxas thanks a lot for you change. Do you anyone in your mind who is familiarized with build stuff and can review it?

**Comment by luxas** [comment]
@piosz I think you and @vishh would be perfect :)

**Comment by piosz** [comment]
@luxas I can take a look from high level Kubernetes perspective, but I don't have enough expertise with go build.

**Comment by vishh** [comment]
I can review the build aspects of this PR.

**Comment by piosz** [comment]
thanks @vishh

**Comment by luxas** [comment]
ping @vishh @piosz 
I hope we'll be able to merge this PR this week so we can release a v1.3 version next week or so

**Comment by luxas** [comment]
I'll rebase when you say you have time to look at it

**Comment by luxas** [comment]
@vishh @piosz Rebased, but this is a WIP item, still testing locally
I guess I'll have it ready tomorrow or on Thursday

Do we want to have influxdb 1.1.1 in the next release? I think that would be great!
In that case, merge #1415 now, before this one, but **do not** push the image, since I would build on top of that PR.

**Comment by luxas** [comment]
@timstclair votes for using the new images that I wrote that are of minimal size, and I vote for that too.

@vishh does it make sense to you to switch to those simpler images as well?
That would also reduce the amount of code that's required.

**Comment by vishh** [comment]
@luxas updating influxdb and grafana would be great. We don't have a
release process for those images since they meant to serve as a reference
implementation only. If the build process is robust, then just documenting
a release process for these images should suffice.

**Comment by luxas** [comment]
Updated this now. This works great on amd64!

Please take a look and merge today so we can test the new release soon.
@vishh @piosz @DirectXMan12

**Comment by luxas** [comment]
@piosz Please merge this now so you feel comfortable with it in time for the release.
It's ready to merge, and simplifies things a lot while adding the manifest list functionality, which means docker will pull the right layers for the right arch automatically.

Ping me if you have any comments, but I've worked on this for so long now so it's quite robust

File: deploy/docker/Dockerfile
Comment (vishh) [inline deploy/docker/Dockerfile]: nit: do retain a maintainer to track the author.

**Review by vishh — APPROVED** [review]

**Comment by vishh** [comment]
Is `make push` the magic command to build and push for all architectures? If yes, I'd suggesting having a `dry run` mode where a real push doesn't happen, just to verify images locally.

File: deploy/docker/Dockerfile
Comment (luxas) [inline deploy/docker/Dockerfile]: The maintainer statement is deprecated https://github.com/docker/docker/pull/25466, also we just did remove all MAINTAINER statements from core k8s dockerfiles

**Comment by luxas** [comment]
@piosz Feel free to merge this.
It worked fine for @vishh 

A `build-all` or something make target that would build for all arches would be an improvement, but that's only a slight improvement.
The default build-only behaviour is still working with `container` or `build`.

I'm excited to get this merged!

**Comment by piosz** [comment]
Thanks for your contribution.

**Comment by luxas** [comment]
@piosz Just make sure you `docker login` to your GCR account before you push everything (i.e. you can't use `gcloud docker` only.

**Pull request merged**
